File: biometric-auth/flutter/crypto_manager.dart
import 'dart:typed_data';
import 'package:encrypt/encrypt.dart' as encrypt;
import 'package:crypto/crypto.dart';
import 'dart:convert';
import 'package:pointycastle/export.dart';

/// Comprehensive Cryptography Manager for Flutter
/// Handles encryption, decryption, key generation, and hashing
class CryptoManager {
  /// Generate AES key
  encrypt.Key generateAESKey() {
    return encrypt.Key.fromSecureRandom(32); // 256-bit
  }

  /// Encrypt data with AES-256-GCM
  EncryptedDataModel encryptAES(Uint8List data, encrypt.Key key) {
    final iv = encrypt.IV.fromSecureRandom(16);
    final encrypter = encrypt.Encrypter(encrypt.AES(key, mode: encrypt.AESMode.gcm));
    
    final encrypted = encrypter.encryptBytes(data, iv: iv);
    
    return EncryptedDataModel(
      ciphertext: encrypted.bytes,
      iv: iv.bytes,
      tag: Uint8List(0), // GCM tag is included in ciphertext
    );
  }

  /// Decrypt data with AES-256-GCM
  Uint8List decryptAES(EncryptedDataModel encryptedData, encrypt.Key key) {
    final iv = encrypt.IV(encryptedData.iv);
    final encrypter = encrypt.Encrypter(encrypt.AES(key, mode: encrypt.AESMode.gcm));
    
    final encrypted = encrypt.Encrypted(encryptedData.ciphertext);
    final decrypted = encrypter.decryptBytes(encrypted, iv: iv);
    
    return Uint8List.fromList(decrypted);
  }

  /// Generate RSA key pair
  RSAKeyPairModel generateRSAKeyPair({int bitLength = 2048}) {
    final keyParams = RSAKeyGeneratorParameters(
      BigInt.parse('65537'), // Public exponent
      bitLength,
      64, // Certainty
    );

    final secureRandom = FortunaRandom();
    final random = SecureRandom('Fortuna')..seed(KeyParameter(
      Uint8List.fromList(List.generate(32, (i) => DateTime.now().millisecondsSinceEpoch % 256))
    ));

    final generator = RSAKeyGenerator()..init(ParametersWithRandom(keyParams, random));
    final keyPair = generator.generateKeyPair();

    final publicKey = keyPair.publicKey as RSAPublicKey;
    final privateKey = keyPair.privateKey as RSAPrivateKey;

    return RSAKeyPairModel(
      publicKey: RSAPublicKeyModel(
        modulus: publicKey.modulus!,
        exponent: publicKey.exponent!,
      ),
      privateKey: RSAPrivateKeyModel(
        modulus: privateKey.modulus!,
        exponent: privateKey.exponent!,
        p: privateKey.p!,
        q: privateKey.q!,
      ),
    );
  }

  /// Encrypt with RSA public key
  Uint8List encryptRSA(Uint8List data, RSAPublicKeyModel publicKey) {
    final cipher = OAEPEncoding(RSAEngine())
      ..init(
        true,
        PublicKeyParameter<RSAPublicKey>(
          RSAPublicKey(publicKey.modulus, publicKey.exponent),
        ),
      );

    return cipher.process(data);
  }

  /// Decrypt with RSA private key
  Uint8List decryptRSA(Uint8List encryptedData, RSAPrivateKeyModel privateKey) {
    final cipher = OAEPEncoding(RSAEngine())
      ..init(
        false,
        PrivateKeyParameter<RSAPrivateKey>(
          RSAPrivateKey(
            privateKey.modulus,
            privateKey.exponent,
            privateKey.p,
            privateKey.q,
          ),
        ),
      );

    return cipher.process(encryptedData);
  }

  /// Derive key from password using PBKDF2
  encrypt.Key deriveKeyFromPassword(
    String password,
    Uint8List salt, {
    int iterations = 100000,
    int keyLength = 32,
  }) {
    final generator = PBKDF2KeyDerivator(HMac(SHA256Digest(), 64))
      ..init(Pbkdf2Parameters(salt, iterations, keyLength));

    final key = generator.process(Uint8List.fromList(password.codeUnits));
    return encrypt.Key(key);
  }

  /// Generate cryptographically secure random salt
  Uint8List generateSalt({int length = 32}) {
    final secureRandom = FortunaRandom();
    final random = SecureRandom('Fortuna')..seed(KeyParameter(
      Uint8List.fromList(List.generate(32, (i) => DateTime.now().millisecondsSinceEpoch % 256))
    ));
    
    return secureRandom.nextBytes(length);
  }

  /// Sign data with RSA private key
  Uint8List signData(Uint8List data, RSAPrivateKeyModel privateKey) {
    final signer = RSASigner(SHA256Digest(), '0609608648016503040201');
    
    signer.init(
      true,
      PrivateKeyParameter<RSAPrivateKey>(
        RSAPrivateKey(
          privateKey.modulus,
          privateKey.exponent,
          privateKey.p,
          privateKey.q,
        ),
      ),
    );

    final signature = signer.generateSignature(data) as RSASignature;
    return signature.bytes;
  }

  /// Verify signature with RSA public key
  bool verifySignature(
    Uint8List data,
    Uint8List signature,
    RSAPublicKeyModel publicKey,
  ) {
    final verifier = RSASigner(SHA256Digest(), '0609608648016503040201');
    
    verifier.init(
      false,
      PublicKeyParameter<RSAPublicKey>(
        RSAPublicKey(publicKey.modulus, publicKey.exponent),
      ),
    );

    try {
      return verifier.verifySignature(data, RSASignature(signature));
    } catch (e) {
      return false;
    }
  }

  /// Hash data with SHA-256
  Uint8List hashSHA256(Uint8List data) {
    return Uint8List.fromList(sha256.convert(data).bytes);
  }

  /// Hash data with SHA-512
  Uint8List hashSHA512(Uint8List data) {
    return Uint8List.fromList(sha512.convert(data).bytes);
  }

  /// HMAC-SHA256
  Uint8List hmacSHA256(Uint8List data, encrypt.Key key) {
    final hmacSha256 = Hmac(sha256, key.bytes);
    return Uint8List.fromList(hmacSha256.convert(data).bytes);
  }

  /// Generate random bytes
  Uint8List generateRandomBytes(int length) {
    final secureRandom = FortunaRandom();
    final seed = Uint8List.fromList(
      List.generate(32, (i) => DateTime.now().millisecondsSinceEpoch % 256),
    );
    secureRandom.seed(KeyParameter(seed));
    
    return secureRandom.nextBytes(length);
  }

  /// Secure wipe byte array
  void secureWipe(Uint8List data) {
    for (int i = 0; i < data.length; i++) {
      data[i] = 0;
    }
  }

  /// Encode to Base64
  String toBase64(Uint8List data) {
    return base64.encode(data);
  }

  /// Decode from Base64
  Uint8List fromBase64(String encoded) {
    return base64.decode(encoded);
  }
}

/// Encrypted data model
class EncryptedDataModel {
  final Uint8List ciphertext;
  final Uint8List iv;
  final Uint8List tag;

  EncryptedDataModel({
    required this.ciphertext,
    required this.iv,
    required this.tag,
  });

  Map<String, dynamic> toJson() => {
        'ciphertext': base64.encode(ciphertext),
        'iv': base64.encode(iv),
        'tag': base64.encode(tag),
      };

  factory EncryptedDataModel.fromJson(Map<String, dynamic> json) {
    return EncryptedDataModel(
      ciphertext: base64.decode(json['ciphertext']),
      iv: base64.decode(json['iv']),
      tag: base64.decode(json['tag']),
    );
  }
}

/// RSA key pair model
class RSAKeyPairModel {
  final RSAPublicKeyModel publicKey;
  final RSAPrivateKeyModel privateKey;

  RSAKeyPairModel({
    required this.publicKey,
    required this.privateKey,
  });
}

/// RSA public key model
class RSAPublicKeyModel {
  final BigInt modulus;
  final BigInt exponent;

  RSAPublicKeyModel({
    required this.modulus,
    required this.exponent,
  });

  String toBase64() {
    final modulusBytes = _bigIntToBytes(modulus);
    final exponentBytes = _bigIntToBytes(exponent);
    final combined = [...modulusBytes, ...exponentBytes];
    return base64.encode(combined);
  }

  Uint8List _bigIntToBytes(BigInt number) {
    final bytes = <int>[];
    var value = number;
    while (value > BigInt.zero) {
      bytes.insert(0, (value & BigInt.from(0xff)).toInt());
      value = value >> 8;
    }
    return Uint8List.fromList(bytes);
  }
}

/// RSA private key model
class RSAPrivateKeyModel {
  final BigInt modulus;
  final BigInt exponent;
  final BigInt p;
  final BigInt q;

  RSAPrivateKeyModel({
    required this.modulus,
    required this.exponent,
    required this.p,
    required this.q,
  });
}

/// Crypto exceptions
class CryptoException implements Exception {
  final String message;
  const CryptoException(this.message);

  @override
  String toString() => 'CryptoException: $message';
}

File: biometric-auth/android-kotlin/CryptoManager.kt
package com.example.security.crypto

import android.os.Build
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import java.security.*
import javax.crypto.*
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * Comprehensive Cryptography Manager
 * Handles encryption, decryption, key generation, and secure storage
 */
class CryptoManager {

    companion object {
        private const val ANDROID_KEYSTORE = "AndroidKeyStore"
        private const val AES_MODE = "AES/GCM/NoPadding"
        private const val RSA_MODE = "RSA/ECB/OAEPWithSHA-256AndMGF1Padding"
        private const val KEY_SIZE = 256
        private const val GCM_TAG_LENGTH = 128
        private const val IV_SIZE = 12
        private const val PBKDF2_ITERATIONS = 100000
    }

    private val keyStore: KeyStore = KeyStore.getInstance(ANDROID_KEYSTORE).apply {
        load(null)
    }

    // ==================== AES Symmetric Encryption ====================

    /**
     * Generate AES key in Android Keystore
     * @param keyAlias Unique identifier for the key
     * @param requireBiometric Whether key requires biometric authentication
     * @param userAuthenticationValiditySeconds How long key is valid after auth (0 = always require)
     */
    fun generateAESKey(
        keyAlias: String,
        requireBiometric: Boolean = false,
        userAuthenticationValiditySeconds: Int = 0,
        useStrongBox: Boolean = false
    ): SecretKey {
        val keyGenerator = KeyGenerator.getInstance(
            KeyProperties.KEY_ALGORITHM_AES,
            ANDROID_KEYSTORE
        )

        val builder = KeyGenParameterSpec.Builder(
            keyAlias,
            KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
        )
            .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
            .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
            .setKeySize(KEY_SIZE)
            .setRandomizedEncryptionRequired(true)

        // Biometric authentication
        if (requireBiometric && Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            builder.setUserAuthenticationParameters(
                userAuthenticationValiditySeconds,
                KeyProperties.AUTH_BIOMETRIC_STRONG
            )
        } else if (requireBiometric && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            builder.setUserAuthenticationRequired(true)
            if (userAuthenticationValiditySeconds > 0) {
                builder.setUserAuthenticationValidityDurationSeconds(
                    userAuthenticationValiditySeconds
                )
            }
        }

        // StrongBox support (Android 9+)
        if (useStrongBox && Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            builder.setIsStrongBoxBacked(true)
        }

        keyGenerator.init(builder.build())
        return keyGenerator.generateKey()
    }

    /**
     * Get AES cipher for encryption
     */
    fun getEncryptCipher(keyAlias: String): Cipher {
        val key = keyStore.getKey(keyAlias, null) as SecretKey
        val cipher = Cipher.getInstance(AES_MODE)
        cipher.init(Cipher.ENCRYPT_MODE, key)
        return cipher
    }

    /**
     * Get AES cipher for decryption
     */
    fun getDecryptCipher(keyAlias: String, iv: ByteArray): Cipher {
        val key = keyStore.getKey(keyAlias, null) as SecretKey
        val cipher = Cipher.getInstance(AES_MODE)
        val spec = GCMParameterSpec(GCM_TAG_LENGTH, iv)
        cipher.init(Cipher.DECRYPT_MODE, key, spec)
        return cipher
    }

    /**
     * Encrypt data with AES-GCM
     */
    fun encryptAES(data: ByteArray, cipher: Cipher): EncryptedData {
        val encryptedBytes = cipher.doFinal(data)
        return EncryptedData(
            ciphertext = encryptedBytes,
            iv = cipher.iv
        )
    }

    /**
     * Decrypt data with AES-GCM
     */
    fun decryptAES(encryptedData: EncryptedData, cipher: Cipher): ByteArray {
        return cipher.doFinal(encryptedData.ciphertext)
    }

    // ==================== RSA Asymmetric Encryption ====================

    /**
     * Generate RSA key pair in Android Keystore
     */
    fun generateRSAKeyPair(
        keyAlias: String,
        keySize: Int = 2048,
        useStrongBox: Boolean = false
    ): KeyPair {
        val keyPairGenerator = KeyPairGenerator.getInstance(
            KeyProperties.KEY_ALGORITHM_RSA,
            ANDROID_KEYSTORE
        )

        val builder = KeyGenParameterSpec.Builder(
            keyAlias,
            KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
        )
            .setKeySize(keySize)
            .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_RSA_OAEP)
            .setDigests(
                KeyProperties.DIGEST_SHA256,
                KeyProperties.DIGEST_SHA512
            )

        if (useStrongBox && Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            builder.setIsStrongBoxBacked(true)
        }

        keyPairGenerator.initialize(builder.build())
        return keyPairGenerator.generateKeyPair()
    }

    /**
     * Encrypt data with RSA public key
     */
    fun encryptRSA(data: ByteArray, publicKey: PublicKey): ByteArray {
        val cipher = Cipher.getInstance(RSA_MODE)
        cipher.init(Cipher.ENCRYPT_MODE, publicKey)
        return cipher.doFinal(data)
    }

    /**
     * Decrypt data with RSA private key
     */
    fun decryptRSA(encryptedData: ByteArray, privateKey: PrivateKey): ByteArray {
        val cipher = Cipher.getInstance(RSA_MODE)
        cipher.init(Cipher.DECRYPT_MODE, privateKey)
        return cipher.doFinal(encryptedData)
    }

    // ==================== Key Derivation ====================

    /**
     * Derive key from password using PBKDF2
     */
    fun deriveKeyFromPassword(
        password: CharArray,
        salt: ByteArray,
        iterations: Int = PBKDF2_ITERATIONS,
        keyLength: Int = KEY_SIZE
    ): SecretKey {
        val spec = PBEKeySpec(password, salt, iterations, keyLength)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        val key = factory.generateSecret(spec)
        return SecretKeySpec(key.encoded, "AES")
    }

    /**
     * Generate cryptographically secure random salt
     */
    fun generateSalt(size: Int = 32): ByteArray {
        val salt = ByteArray(size)
        SecureRandom().nextBytes(salt)
        return salt
    }

    // ==================== Digital Signatures ====================

    /**
     * Generate key pair for signing
     */
    fun generateSigningKeyPair(
        keyAlias: String,
        useECC: Boolean = true
    ): KeyPair {
        val algorithm = if (useECC) {
            KeyProperties.KEY_ALGORITHM_EC
        } else {
            KeyProperties.KEY_ALGORITHM_RSA
        }

        val keyPairGenerator = KeyPairGenerator.getInstance(
            algorithm,
            ANDROID_KEYSTORE
        )

        val builder = KeyGenParameterSpec.Builder(
            keyAlias,
            KeyProperties.PURPOSE_SIGN or KeyProperties.PURPOSE_VERIFY
        )
            .setDigests(KeyProperties.DIGEST_SHA256, KeyProperties.DIGEST_SHA512)

        if (useECC) {
            // Use P-256 curve for ECC
        } else {
            builder.setKeySize(2048)
                .setSignaturePaddings(KeyProperties.SIGNATURE_PADDING_RSA_PSS)
        }

        keyPairGenerator.initialize(builder.build())
        return keyPairGenerator.generateKeyPair()
    }

    /**
     * Sign data
     */
    fun signData(data: ByteArray, privateKey: PrivateKey, useECC: Boolean = true): ByteArray {
        val algorithm = if (useECC) "SHA256withECDSA" else "SHA256withRSA/PSS"
        val signature = Signature.getInstance(algorithm)
        signature.initSign(privateKey)
        signature.update(data)
        return signature.sign()
    }

    /**
     * Verify signature
     */
    fun verifySignature(
        data: ByteArray,
        signatureBytes: ByteArray,
        publicKey: PublicKey,
        useECC: Boolean = true
    ): Boolean {
        val algorithm = if (useECC) "SHA256withECDSA" else "SHA256withRSA/PSS"
        val signature = Signature.getInstance(algorithm)
        signature.initVerify(publicKey)
        signature.update(data)
        return signature.verify(signatureBytes)
    }

    // ==================== Hashing ====================

    /**
     * Hash data with SHA-256
     */
    fun hashSHA256(data: ByteArray): ByteArray {
        val digest = MessageDigest.getInstance("SHA-256")
        return digest.digest(data)
    }

    /**
     * Hash data with SHA-512
     */
    fun hashSHA512(data: ByteArray): ByteArray {
        val digest = MessageDigest.getInstance("SHA-512")
        return digest.digest(data)
    }

    /**
     * HMAC-SHA256
     */
    fun hmacSHA256(data: ByteArray, key: SecretKey): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(key)
        return mac.doFinal(data)
    }

    // ==================== Key Management ====================

    /**
     * Check if key exists
     */
    fun keyExists(keyAlias: String): Boolean {
        return keyStore.containsAlias(keyAlias)
    }

    /**
     * Delete key
     */
    fun deleteKey(keyAlias: String) {
        if (keyExists(keyAlias)) {
            keyStore.deleteEntry(keyAlias)
        }
    }

    /**
     * Get all key aliases
     */
    fun getAllKeyAliases(): List<String> {
        return keyStore.aliases().toList()
    }

    /**
     * Get public key
     */
    fun getPublicKey(keyAlias: String): PublicKey? {
        return (keyStore.getEntry(keyAlias, null) as? KeyStore.PrivateKeyEntry)?.certificate?.publicKey
    }

    /**
     * Get private key
     */
    fun getPrivateKey(keyAlias: String): PrivateKey? {
        return keyStore.getKey(keyAlias, null) as? PrivateKey
    }

    // ==================== Utility Functions ====================

    /**
     * Secure wipe byte array
     */
    fun secureWipe(data: ByteArray) {
        data.fill(0)
    }

    /**
     * Generate random bytes
     */
    fun generateRandomBytes(size: Int): ByteArray {
        val bytes = ByteArray(size)
        SecureRandom().nextBytes(bytes)
        return bytes
    }
}

/**
 * Encrypted data container
 */
data class EncryptedData(
    val ciphertext: ByteArray,
    val iv: ByteArray,
    val authTag: ByteArray? = null
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as EncryptedData

        if (!ciphertext.contentEquals(other.ciphertext)) return false
        if (!iv.contentEquals(other.iv)) return false
        if (authTag != null) {
            if (other.authTag == null) return false
            if (!authTag.contentEquals(other.authTag)) return false
        } else if (other.authTag != null) return false

        return true
    }

    override fun hashCode(): Int {
        var result = ciphertext.contentHashCode()
        result = 31 * result + iv.contentHashCode()
        result = 31 * result + (authTag?.contentHashCode() ?: 0)
        return result
    }
}
